File: ios/Keyboard/UI/KeyboardLayoutView.swift
import UIKit

enum ShiftState { case off, oneShot, locked }

/// The key plane: rows of `KeyButton`s laid out by proportional weights (so the
/// same model fits any iPhone/iPad width and orientation). Reports taps as `KeyCap`s.
/// A `.spacer` spec reserves width without a button, used to center short rows.
final class KeyboardLayoutView: UIView {
    var onKey: ((KeyCap) -> Void)?
    weak var popupHost: UIView?

    private struct RowItem {
        let spec: KeySpec
        let button: KeyButton?   // nil for spacers
    }
    private var rows: [[RowItem]] = []
    private var theme: KeyboardTheme
    private let hGap: CGFloat
    private let vGap: CGFloat = 8
    private let sideMargin: CGFloat
    private let topMargin: CGFloat = 5

    /// Localized label for the return key (varies by the host's returnKeyType).
    var returnLabel = "换行" {
        didSet { forEachButton { if $0.spec.cap == .ret { $0.displayText = returnLabel } } }
    }

    init(theme: KeyboardTheme, idiom: KeyboardIdiom) {
        self.theme = theme
        self.hGap = idiom == .pad ? 7 : 5
        self.sideMargin = idiom == .pad ? 5 : 3
        super.init(frame: .zero)
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    private func forEachButton(_ body: (KeyButton) -> Void) {
        for row in rows { for item in row { if let b = item.button { body(b) } } }
    }

    func setRows(_ specRows: [[KeySpec]]) {
        forEachButton { $0.removeFromSuperview() }
        rows = specRows.map { row in
            row.map { spec -> RowItem in
                guard spec.cap != .spacer else { return RowItem(spec: spec, button: nil) }
                let b = KeyButton(spec: spec, theme: theme)
                b.popupHost = popupHost
                if spec.cap == .ret { b.displayText = returnLabel }
                b.onTap = { [weak self] cap in self?.onKey?(cap) }
                addSubview(b)
                return RowItem(spec: spec, button: b)
            }
        }
        setNeedsLayout()
    }

    func apply(theme: KeyboardTheme) {
        self.theme = theme
        forEachButton { $0.apply(theme: theme) }
    }

    func updateShift(_ state: ShiftState) {
        forEachButton { b in
            switch b.spec.cap {
            case .char(let byte) where (0x61...0x7A).contains(byte):
                let lower = String(UnicodeScalar(byte))
                b.displayText = state == .off ? lower : lower.uppercased()
            case .shift:
                b.displayText = state == .locked ? "⇪" : "⇧"
                b.isAccented = state != .off
            default:
                break
            }
        }
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        let rowCount = rows.count
        guard rowCount > 0 else { return }
        let totalV = bounds.height - 2 * topMargin - CGFloat(rowCount - 1) * vGap
        let rowHeight = totalV / CGFloat(rowCount)

        var y = topMargin
        for row in rows {
            let sumWeights = row.reduce(0) { $0 + $1.spec.weight }
            let availW = bounds.width - 2 * sideMargin - CGFloat(row.count - 1) * hGap
            var x = sideMargin
            for item in row {
                let w = availW * (item.spec.weight / sumWeights)
                item.button?.frame = CGRect(x: x, y: y, width: w, height: rowHeight)
                x += w + hGap
            }
            y += rowHeight + vGap
        }
    }
}

File: android/app/src/main/java/com/hronro/imejd/ui/KeyboardLayoutView.kt
// Port of ios/Keyboard/UI/KeyboardLayoutView.swift. The key plane: rows of
// KeyButtons laid out by proportional weights (so one model fits any width and
// orientation). A `.Spacer` spec reserves width without a button, to center rows.
package com.hronro.imejd.ui

import android.annotation.SuppressLint
import android.content.Context
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import com.hronro.imejd.R
import kotlin.math.roundToInt

enum class ShiftState { OFF, ONE_SHOT, LOCKED }

@SuppressLint("ViewConstructor")
class KeyboardLayoutView(
    context: Context,
    private var theme: KeyboardTheme,
    idiom: KeyboardIdiom,
) : ViewGroup(context) {

    var onKey: ((KeyCap) -> Unit)? = null

    /** Forwarded KeyButton preview events; the owner routes them to the balloon layer. */
    var onKeyPreview: ((KeyButton, KeyPreviewEvent) -> Unit)? = null

    var returnLabel: String = "换行"
        set(value) {
            field = value
            forEachButton { if (it.spec.cap is KeyCap.Return) it.displayText = value }
        }

    private class RowItem(val spec: KeySpec, val button: KeyButton?)
    private var rows: List<List<RowItem>> = emptyList()
    private var shift: ShiftState = ShiftState.OFF

    private val density = context.resources.displayMetrics.density
    private val hGap = (if (idiom == KeyboardIdiom.PAD) 7f else 5f) * density
    private val vGap = 8f * density
    private val sideMargin = (if (idiom == KeyboardIdiom.PAD) 5f else 3f) * density
    private val topMargin = 5f * density

    init { clipChildren = false }

    private inline fun forEachButton(body: (KeyButton) -> Unit) {
        for (row in rows) for (item in row) item.button?.let(body)
    }

    fun setRows(specRows: List<List<KeySpec>>) {
        removeAllViews()
        rows = specRows.map { row ->
            row.map { spec ->
                if (spec.cap is KeyCap.Spacer) {
                    RowItem(spec, null)
                } else {
                    val b = KeyButton(context, spec, theme)
                    if (spec.cap is KeyCap.Return) b.displayText = returnLabel
                    b.onTap = { cap -> onKey?.invoke(cap) }
                    b.onPreview = { key, event -> onKeyPreview?.invoke(key, event) }
                    addView(b)
                    RowItem(spec, b)
                }
            }
        }
        updateShift(shift)
        requestLayout()
    }

    fun apply(theme: KeyboardTheme) {
        this.theme = theme
        forEachButton { it.apply(theme) }
    }

    fun updateShift(state: ShiftState) {
        shift = state
        forEachButton { b ->
            when (val cap = b.spec.cap) {
                is KeyCap.Char -> {
                    val byte = cap.byte.toInt() and 0xFF
                    if (byte in 0x61..0x7A) {
                        val lower = byte.toChar().toString()
                        b.displayText = if (state == ShiftState.OFF) lower else lower.uppercase()
                    }
                }
                KeyCap.Shift -> {
                    b.icon = ContextCompat.getDrawable(
                        context,
                        if (state == ShiftState.LOCKED) R.drawable.ic_caps_lock else R.drawable.ic_shift,
                    )
                    b.isAccented = state != ShiftState.OFF
                }
                else -> {}
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val w = MeasureSpec.getSize(widthMeasureSpec)
        val h = MeasureSpec.getSize(heightMeasureSpec)
        setMeasuredDimension(w, h)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val rowCount = rows.size
        if (rowCount == 0) return
        val width = (r - l).toFloat()
        val height = (b - t).toFloat()
        val totalV = height - 2 * topMargin - (rowCount - 1) * vGap
        val rowHeight = totalV / rowCount

        var y = topMargin
        for (row in rows) {
            val sumWeights = row.fold(0f) { acc, item -> acc + item.spec.weight }
            val availW = width - 2 * sideMargin - (row.size - 1) * hGap
            var x = sideMargin
            for (item in row) {
                val cw = availW * (item.spec.weight / sumWeights)
                item.button?.let { btn ->
                    val left = x.roundToInt()
                    val top = y.roundToInt()
                    btn.measure(
                        MeasureSpec.makeMeasureSpec(cw.roundToInt(), MeasureSpec.EXACTLY),
                        MeasureSpec.makeMeasureSpec(rowHeight.roundToInt(), MeasureSpec.EXACTLY),
                    )
                    btn.layout(left, top, (x + cw).roundToInt(), (y + rowHeight).roundToInt())
                }
                x += cw + hGap
            }
            y += rowHeight + vGap
        }
    }
}
